[Merged by Bors] - feat(algebra/*,linear_algebra/basic,ring_theory/ideal): lemmas about span of finite subsets and nontrivial maximal ideals

laughinggas opened the pull request:
---
<!-- put comments you want to keep out of the PR commit here.
If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]
-->
This is to add lemmas regarding span of submodules, and finite sets contained in them. There is also a lemma stating that maximal ideals are nontrivial if the ring is not a field. @Vierkantor helped me refine and make the proofs readable + elegant, and created the lemma `mem_span_mul_finite_of_mem_mul`.

Vierkantor committed:
Rebase upon latest mathlib, to avoid merge conflicts later on.

This is to add lemmas regarding span of submodules, and finite sets contained in them. There is also a lemma stating that maximal ideals are nontrivial if the ring is not a field. @Vierkantor helped me refine and make the proofs readable + elegant, and created the lemma `mem_span_mul_finite_of_mem_mul`.

Co-Authored-By: Ashvni <ashvni.n@gmail.com>

Vierkantor committed:
Fix overlong line

jcommelin commented on src/algebra/algebra/operations.lean:
```diff
@@ -164,6 +164,44 @@ calc map f.to_linear_map (M * N)
+lemma mem_span_mul_finite_of_mem_span_mul {S : set A} {S' : set A} {x : A}
+  (hx : x ∈ span R (S * S')) :
+  ∃ (T T' : finset A), ↑T ⊆ S ∧ ↑T' ⊆ S' ∧ x ∈ span R (T * T' : set A) :=
+begin
```
I feel like this should combine `mem_span_finite_of_mem_span` that you add below, and another lemma:

If `U` is a finite subset of `S * S'`, then there exist finite `T \sub S` and `T' \sub S'` such that `U \sub T * T'`. There will probably be some rough edges that you need to smoothen.

Combining those two lemmas should give a more direct proof of this lemma, instead of repeating the `span_induction`.

laughinggas replied:
Apologies for the delay. I have made changes. The proof isn't shorter, but the added lemma might be useful.

laughinggas committed:
Proof change

Changed proof of `mem_span_mul_finite_of_mem_span_mul` and added `finset.subset_mul`.

laughinggas committed:
changed proof

laughinggas committed:
Merge branch 'submodule-related' of https://github.com/leanprover-community/mathlib into submodule-related

jcommelin commented on src/algebra/algebra/operations.lean:
```diff
@@ -164,6 +164,46 @@ calc map f.to_linear_map (M * N)
+namespace finset
+/-- A finite set `U` contained in the product of two sets `S * S'` is also contained in the product
+of two finite sets `T * T' ⊆ S * S'`. -/
+lemma subset_mul {M : Type*} [monoid M] {S : set M} {S' : set M} {U : finset M} (f : ↑U ⊆ S * S') :
```
I think this lemma should now move to a file about pointwise multiplication (`algebra/pointwise1, I think?)

jcommelin replied:
Also, if you add the `@[to_additive]` attribute, you'll have `subset_add` automatically.

laughinggas replied:
I see, thank you! I have moved the lemma.

laughinggas committed:
Moved lemma to correct file

jcommelin commented on src/ring_theory/ideal/basic.lean:
```diff
@@ -576,6 +576,18 @@ not_is_field_iff_exists_ideal_bot_lt_and_lt_top.trans
+lemma ne_bot_of_is_maximal_of_not_is_field [nontrivial R] {M : ideal R} (max : M.is_maximal)
+  (not_field : ¬ is_field R) : M ≠ ⊥ :=
+begin
+  rintros h,
+  rw h at max,
```
```suggestion
  rintro rfl,
```

jcommelin commented on src/ring_theory/ideal/basic.lean:
```diff
@@ -576,6 +576,18 @@ not_is_field_iff_exists_ideal_bot_lt_and_lt_top.trans
+  cases max with h1 h2,
+  obtain ⟨I, hIbot, hItop⟩ := not_is_field_iff_exists_ideal_bot_lt_and_lt_top.mp not_field,
+  specialize h2 I hIbot,
+  exact ne_of_lt hItop h2,
```
```suggestion
  exact ne_of_lt hItop (h2 I hIbot),
```

jcommelin commented on src/algebra/pointwise.lean:
```diff
@@ -410,6 +410,28 @@ by { convert finset.card_image_le, rw [finset.card_product, mul_comm] }
+  use [insert x T, insert y T'], simp only [finset.coe_insert],
+  repeat {rw [set.insert_subset], }, use [hx, hS, hy, hS'],
+  refine finset.insert_subset.mpr _,
+  split, rw finset.mem_mul, use [x,y],
```
This `split` creates two goals. Please put `{ }` around the goals to indicate this.

laughinggas replied:
I am forgetful about that, thank you for pointing it out! I have made the changes.

jcommelin reviewed (commented):
Sorry. I lost track of this PR. But it looks good! A couple more comments, and then it's ready to go.

laughinggas committed:
Update src/ring_theory/ideal/basic.lean

Co-authored-by: Johan Commelin <johan@commelin.net>

laughinggas committed:
Update src/ring_theory/ideal/basic.lean

Co-authored-by: Johan Commelin <johan@commelin.net>

laughinggas committed:
inserted brackets for split

laughinggas committed:
changing rintros rfl to rintros rw

jcommelin reviewed (approved):
Thanks :tada:

bors merge

bryangingechen committed:
formatting

bryangingechen committed:
Merge remote-tracking branch 'community/master' into submodule-related

bryangingechen committed:
put open_locale classical in a section

bryangingechen committed:
Merge remote-tracking branch 'community/master' into submodule-related
